lib/stage.ts

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2019 Balena Ltd
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+   http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import * as _ from 'lodash';
+import * as path from 'path';
+
+import ActionGroup, {
+	Command,
+	fileMatchesForActionGroup,
+	LocalCopy,
+	LocalDependentActionGroup,
+	StageDependentActionGroup,
+} from './action-group';
+import { DockerfileParseError } from './errors';
+
+export class Stage {
+	public name: string;
+	public index: number;
+	public dependentOnStages: number[] = [];
+	public isLast: boolean = false;
+	public actionGroups: ActionGroup[] = [
+		{
+			dependentOnStage: false,
+			commands: [],
+			copies: [],
+			workdir: '/',
+		},
+	];
+
+	/*
+		This var is used to detect whether we create a new action group,
+		as we can collect as many copies together into the same action group,
+		but to ensure we run the commands in the correct order, a new copy forces
+		a new group
+	*/
+	private lastStepWasCopy: boolean = false;
+	private lastWorkdir: string = '/';
+	private ungroupedCommands: Command[] = [];
+
+	public constructor(idx: number, name?: string) {
+		this.index = idx;
+		this.name = name != null ? name : idx.toString();
+	}
+
+	public addLocalCopyStep(args: string[]) {
+		const lastActionGroup = _.last(
+			this.actionGroups,
+		) as LocalDependentActionGroup;
```

Comment:
I think you can get rid of this cast with
```ts
if (actionGroup == null) {
	throw new Error(...)
}
if (actionGroup.dependentOnStage === true) {
	throw new Error(...)
}
```
as that should narrow the type correctly and guarantee the type is what we expect